Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs

```diff
@@ -0,0 +1,367 @@
+    /// converting something into it's single juice form. E.g, grind an apple and get the nutriment and sugar
+    /// it contained, juice an apple and get "apple juice".
+    /// </summary>
+    public class ReagentGrinderComponent : SharedReagentGrinderComponent, IActivate, IInteractUsing
+    {
+        [Dependency] private readonly IEntityManager _entityManager = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly IRobustRandom _random = default!;
+
+
+        private AudioSystem _audioSystem = default!;
+
+        [ViewVariables] private ContainerSlot _beakerContainer = default!;
+
+        /// <summary>
+        /// Can be null since we won't always have a beaker in the grinder.
+        /// </summary>
+        [ViewVariables] private SolutionContainerComponent? _heldBeaker = default!;
+
+        /// <summary>
+        /// Contains the things that are going to be ground or juiced.
+        /// </summary>
+        [ViewVariables] private Container _chamber = default!;
+
+        [ViewVariables] private bool ChamberEmpty => _chamber.ContainedEntities.Count <= 0;
+        [ViewVariables] private bool HasBeaker => _beakerContainer.ContainedEntity != null;
+
+        [ViewVariables] private BoundUserInterface? UserInterface => Owner.GetUIOrNull(ReagentGrinderUiKey.Key);
+
+
+        private bool Powered => !Owner.TryGetComponent(out PowerReceiverComponent? receiver) || receiver.Powered;
+
+        /// <summary>
+        /// Should the BoundUI be told to update?
+        /// </summary>
+        private bool _dirty = true;
+        /// <summary>
+        /// Is the machine actively doing something and can't be used right now?
+        /// </summary>
+        private bool _busy = false;
+
+
+        //YAML serialization vars
+        private int _storageCap = 16;
+        private int _workTime = 3500; //3.5 seconds, completely arbitrary for now.
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);    
+            serializer.DataField(ref _storageCap, "chamber_capacity", 16);
+            serializer.DataField(ref _workTime, "time", 3500);
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            //A slot for the beaker where the grounds/juices will go.
+            _beakerContainer =
+                ContainerManagerComponent.Ensure<ContainerSlot>($"{Name}-reagentContainerContainer", Owner);
+
+            //A container for the things that WILL be ground/juiced. Useful for ejecting them instead of deleting them from the hands of the user.
+            _chamber =
+                ContainerManagerComponent.Ensure<Container>($"{Name}-entityContainerContainer", Owner);
+
+            if (UserInterface != null)
+            {
+                UserInterface.OnReceiveMessage += UserInterfaceOnReceiveMessage;
```

Comment:
Unsubscribe though I think there's an issue to refactor it to not need this.

From memory if you don't it never gets GC'd.